[Feature] Synchrofact Detection

Kleinjohann opened the pull request:
This is a method to detect highly synchronous spikes (at the level of sampling rate precision with an option to extend this to jittered synchrony) and annotate or optionally remove them. 

The detection of complexity intervals (time intervals with at least `n` spikes separated by strictly fewer than `spread - 1` empty bins) is particularly suited for artefact detection and analysis, but can be used to get a (time-resolved) overview of complexities at any time-scale. This is different from `statistics.time_histogram` and `statistics.complexity_pdf` since it uses the amount of empty bins separating spikes, not a simple bincount and not a moving window with a fixed width while still providing time-resolved output.

I started from an institute-internal script and adapted it to use a `BinnedSpikeTrain` in the histogramming procedure and I adapted a set of basic tests I wrote for said internal script.

pep8speaks commented:
Hello @Kleinjohann! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`elephant/spike_train_processing.py`](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/spike_train_processing.py):

> [Line 261:1](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/spike_train_processing.py#L261): [W391](https://duckduckgo.com/?q=pep8%20W391) blank line at end of file

* In the file [`elephant/test/test_spike_train_processing.py`](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py):

> [Line 19:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L19): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 57:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L57): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 67:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 68:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L68): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 69:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L69): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 70:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L70): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 72:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L72): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 73:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L73): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)
> [Line 75:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L75): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 81:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L81): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 82:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L82): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 83:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L83): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 84:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L84): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 89:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L89): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 96:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L96): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 115:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L115): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 116:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L116): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 117:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L117): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 118:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L118): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 120:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L120): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 121:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L121): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 123:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L123): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 129:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L129): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 130:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L130): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 131:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L131): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 132:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L132): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 135:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L135): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 139:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L139): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (85 > 79 characters)
> [Line 160:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L160): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 161:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L161): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 162:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L162): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 163:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L163): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 170:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L170): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 176:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L176): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 177:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L177): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 178:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L178): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 179:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L179): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 185:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L185): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 190:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L190): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 207:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L207): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 208:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L208): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 209:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L209): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 210:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L210): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 217:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L217): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 223:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L223): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 224:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L224): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 225:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L225): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 226:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L226): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 232:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L232): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 237:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L237): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 251:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L251): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 254:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L254): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 257:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L257): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 258:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L258): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 259:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L259): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 260:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L260): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 267:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L267): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 273:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L273): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 274:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L274): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 275:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L275): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 276:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L276): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 282:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L282): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 287:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L287): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 308:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L308): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (83 > 79 characters)
> [Line 310:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L310): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (82 > 79 characters)
> [Line 315:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L315): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (81 > 79 characters)
> [Line 324:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L324): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 325:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L325): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 326:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L326): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 327:51](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L327): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 334:80](https://github.com/NeuralEnsemble/elephant/blob/4ab6ff85844b7e4eb2fccb0217c1fdcf1b3c0f12/elephant/test/test_spike_train_processing.py#L334): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)

morales-gregorio commented:
Solves #326 

From my side this PR is ready for review.

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    return complexity_epoch
+
+
+def precise_complexity_histogram(spiketrains, **kwargs):
```
All 3 functions in this module share the very same arguments - sampling_rate and spread. Consider switching your API to OOP. I can easily imagine a new functionality is being added with similar API in this module. Then you need to double the description and keep track of passing the arguments here and there.

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+
+    Examples
+    --------
+    Here the behavior of
+    `elephant.spike_train_processing.precise_complexity_intervals` is shown, by
+    applying the function to some sample spiketrains.
```
The description of an example is optional and not required if there is nothing more to say than the code speaks by itself.

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    bst = conv.BinnedSpikeTrain(spiketrains,
+                                binsize=bin_size)
+
+    bincount = np.array(bst.to_sparse_array().sum(axis=0)).squeeze()
```
is this the number of spikes?
if so, use `bst.get_num_of_spikes(axis=0)`

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+        bin_indices = np.nonzero(bincount)[0]
+        complexities = bincount[bin_indices]
+        left_edges = bst.bin_edges[bin_indices]
+        right_edges = bst.bin_edges[bin_indices + 1]
```
are you sure that the right edges always exist? I mean, cannot `bin_edges[bin_indices + 1]` be outside of the range?

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+        left_edges = bst.bin_edges[bin_indices]
+        right_edges = bst.bin_edges[bin_indices + 1]
+    else:
+        i = 0
```
give a meaningful name to `i`
is this a spike train id?
at least drop a comment

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+                last_nonzero_index = 0
+                current_window = bincount[i:i + spread + 1]
+                window_sum = current_window.sum()
+                while window_sum > last_window_sum:
```
there is a heuristic: if you use >3 levels deep abstraction (in your case it's else-while-else-while=4), consider moving some part to a separate function. It's hard to read.
Switching to OOP could also help to split the logic of this 4-layers deep block into smaller meaningful pieces.

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+
+        # we dropped units above, because neither concatenate nor append works
+        # with arrays of quantities
+        left_edges *= bst.bin_edges.units
```
I wonder how this works... in-place operation cannot change the original type of an object, and you're changing a list to Quantity. An error should be raised here, should it not?

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    right_edges -= bin_shift
+
+    # ensure that an epoch does not start before the minimum t_start
+    min_t_start = min([st.t_start for st in spiketrains])
```
no need to pass `[` and `]`, `min()` works with iterators
so a simpler and valid version is `min(st.t_start for st in spiketrains)`

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    """
+    Given a list of neo.Spiketrain objects, calculate the number of synchronous
+    spikes found and optionally delete or extract them from the given list
+    *in-place*.
```
Should be "a list of `neo.Spiketrain` objects"

dizcza replied:
You're definitely giving Alex a hard time!
:D
I didn't dare to comment on such things. I think they are negligible anyhow, especially since the API is not yet confirmed.

kohlerca replied:
:D. Just to flag things in advance. Should the API change, the documentation may just be moved around.

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+
+    Parameters
+    ----------
+    spiketrains : list of neo.SpikeTrains
```
Should be "list of neo.SpikeTrain"

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    Parameters
+    ----------
+    spiketrains : list of neo.SpikeTrains
+        a list of neo.SpikeTrains objects. These spike trains should have been
```
Should be neo.SpikeTrain enclosed in backticks

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    sampling_rate : pq.Quantity
+        Sampling rate of the spike trains. The spike trains are binned with
+        bin_size = 1 / `sampling_rate`.
+    spread : int
```
optional

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+            spikes and there are definitely more efficient ways of doing this)
+          * `deletion_threshold <= 0` leads to a ValueError.
+        Default: None
+    invert_delete : bool
```
optional

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+
+    Parameters
+    ----------
+    spiketrains : list of neo.SpikeTrains
```
Should be list of neo.SpikeTrain

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    Parameters
+    ----------
+    spiketrains : list of neo.SpikeTrains
+        a list of neo.SpikeTrains objects. These spike trains should have been
```
Should be neo.SpikeTrain enclosed in backticks

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    [1.5 5. ] ms
+
+    """
+    _check_consistency_of_spiketrainlist(spiketrains)
```
Contrary to the function before, this is not checking if `spiketrains` is a list

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+
+    Parameters
+    ----------
+    spiketrains : list of neo.SpikeTrains
```
Should be list of neo.SpikeTrain

kohlerca commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,317 @@
+    complexity_histogram : np.ndarray
+        A histogram of complexities. `complexity_histogram[i]` corresponds
+        to the number of events of complexity `i` for `i > 0`.
+    """
```
Missing See Also if arguments are related to another function

morales-gregorio commented:
Hey!

We have finally managed to commit the class structure for the `complexity` and the `synchrotool`.

The documentation of `complexity` is in principle complete. Although we are seeing some weird sections emerge (unexpected methods and attributes, which do not have docstrings per se).

We still do not  have the working documentation for `synchrotool`, and we still need to think if we want to inherit the documentation from `complexity` or not.  We will write down a better version of that documentation and reach out again.

Let us know if you want to discuss this in person at some meeting :)

Cheers,
Aitor and Alex

dizcza commented on elephant/statistics.py:
```diff
@@ -799,23 +809,416 @@ def complexity_pdf(spiketrains, binsize):
+    return complexity_obj.pdf
+
+
+class complexity:
```
Is there a particular reason why you called the class name lowercase? Python standards tell using CamelCase naming for classes. Same for `Synchrotool` (or `SynchroTool`) class.

dizcza commented on elephant/statistics.py:
```diff
@@ -799,23 +809,416 @@ def complexity_pdf(spiketrains, binsize):
+            self.epoch = self._epoch_with_spread()
+            self.time_histogram, self.histogram = self._histogram_with_spread()
+
+    @property
```
Why property?
It is not a _characteristic_ of an object. PDF should be a function.

dizcza commented on elephant/statistics.py:
```diff
@@ -799,23 +809,416 @@ def complexity_pdf(spiketrains, binsize):
+        min_t_start = min([st.t_start for st in self.input_spiketrains])
+        if left_edges[0] < min_t_start:
+            left_edges[0] = min_t_start
+            durations[0] -= bin_shift
```
`bin_shift` is not defined when self.sampling_rate if False.

Kleinjohann replied:
True, I fixed it.

dizcza commented on elephant/statistics.py:
```diff
@@ -799,23 +809,416 @@ def complexity_pdf(spiketrains, binsize):
+        else:
+            bincount = np.array(bst.to_sparse_array().sum(axis=0)).squeeze()
+
+        i = 0
+        complexities = []
+        left_edges = []
+        right_edges = []
+        while i < len(bincount):
+            current_bincount = bincount[i]
+            if current_bincount == 0:
+                i += 1
+            else:
+                last_window_sum = current_bincount
+                last_nonzero_index = 0
+                current_window = bincount[i:i + self.spread + 1]
+                window_sum = current_window.sum()
+                while window_sum > last_window_sum:
+                    last_nonzero_index = np.nonzero(current_window)[0][-1]
+                    current_window = bincount[i:
+                                              i + last_nonzero_index
+                                              + self.spread + 1]
+                    last_window_sum = window_sum
+                    window_sum = current_window.sum()
+                complexities.append(window_sum)
+                left_edges.append(
+                    bst.bin_edges[i].magnitude.item())
+                right_edges.append(
+                    bst.bin_edges[
+                        i + last_nonzero_index + 1
+                    ].magnitude.item())
+                i += last_nonzero_index + 1
```
this block will be the bottleneck in understanding the functionality in the future.
I do see that you have spent time thinking on appropriate names that could "tell a story" and yet...
I don't have an alternative suggestion. Keeping it in the current state for now.

dizcza commented on elephant/utils.py:
```diff
@@ -17,3 +19,28 @@ def is_binary(array):
     return ((array == 0) | (array == 1)).all()
+
+
+def _check_consistency_of_spiketrainlist(spiketrains,
```
On second thought, I agree: better to keep it private.

dizcza commented on elephant/statistics.py:
```diff
@@ -799,23 +809,416 @@ def complexity_pdf(spiketrains, binsize):
+            regardless of the number of spikes.
+          * If ``binary = False`` : Number of neurons and spikes per neurons
+            in each bin.
+    complexity_histogram : np.ndarray
```
I don't see any references to `complexity_histogram` in either of the classes.

dizcza commented on elephant/spike_train_processing.py:
```diff
@@ -0,0 +1,246 @@
+
+        self.annotated = True
+
+    def _get_index(self, lst, obj):
```
no need for this method
it's equivalent to `lst.index(obj)` with catching the ValueError

Kleinjohann replied:
```
import neo
import numpy as np
import quantities as pq

st1 = neo.SpikeTrain(np.arange(5)*pq.s, t_stop=5*pq.s)
st2 = neo.SpikeTrain(np.arange(5)*pq.s, t_stop=5*pq.s)

l = [st1, st2]
l.index(st2)
```
causes
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
So `lst.index(obj)` seems to check for equality rather than object identity.

dizcza replied:
You're right, I didn't know that. It's even written in the python docs 

> Return zero-based index in the list of the first item whose value is equal to x.

Then I see the need for the method.

dizcza commented:
I find this PR ready for the final review and merge.
Since it depends on another ongoing PR (in terms of the file location of your methods) about the Spike Contrast method, let's wait for the latter to be merged first.

dizcza commented:
Ready for the final review.

kohlerca commented on elephant/spike_train_synchrony.py:
```diff
@@ -218,3 +228,146 @@ def spike_contrast(spiketrains, t_start=None, t_stop=None,
+            Determines whether the modification are made in place
+            on ``self.input_spiketrains``.
+            Default: False
+        mode : bool
```
mode : {'delete', 'extract'}

Kleinjohann replied:
Done

kohlerca commented on elephant/statistics.py:
```diff
@@ -974,23 +977,414 @@ def complexity_pdf(spiketrains, bin_size):
+          * If both are given then `bin_size` is used.
+        Default: None
+    binary : bool, optional
+          * If `True` then the time histograms will be binary.
```
No backticks in the boolean values True/False

kohlerca commented on elephant/statistics.py:
```diff
@@ -974,23 +977,414 @@ def complexity_pdf(spiketrains, bin_size):
+    bin_size : pq.Quantity or None, optional
+        Width of the histogram's time bins with units of time.
+        The user must specify the `bin_size` or the `sampling_rate`.
+          * If no `bin_size` is specified and the `sampling_rate` is available
```
If None and the `sampling_rate`...

Kleinjohann replied:
Done

kohlerca commented on elephant/statistics.py:
```diff
@@ -974,23 +977,414 @@ def complexity_pdf(spiketrains, bin_size):
+    ----------
+    spiketrains : list of neo.SpikeTrain
+        Spike trains with a common time axis (same `t_start` and `t_stop`)
+    sampling_rate : pq.Quantity or None, optional
```
Missing description of behavior when None

Kleinjohann replied:
Done

kohlerca commented on elephant/statistics.py:
```diff
@@ -974,23 +977,414 @@ def complexity_pdf(spiketrains, bin_size):
+          * ``spread = n`` corresponds to counting spikes separated by exactly
+            or less than `n - 1` empty bins.
+        Default: 0
+    tolerance : float, optional
```
If None can be passed, the information and behavior should be added here. If warnings are raised, then a section Warns can be added

Kleinjohann replied:
Done

kohlerca commented on elephant/spike_train_synchrony.py:
```diff
@@ -218,3 +228,146 @@ def spike_contrast(spiketrains, t_start=None, t_stop=None,
+              * `deletion_threshold <= 1` leads to a ValueError, since this
+              would delete/extract all spikes and there are definitely more
+              efficient ways of doing so.
+        in_place : bool
```
bool, optional

Kleinjohann replied:
Done

kohlerca commented on elephant/spike_train_synchrony.py:
```diff
@@ -218,3 +228,146 @@ def spike_contrast(spiketrains, t_start=None, t_stop=None,
+                                          tolerance=tolerance)
+
+    def delete_synchrofacts(self, threshold, in_place=False, mode='delete'):
+        """
```
Missing Raises section

Kleinjohann replied:
Done

JuliaSprenger commented on elephant/test/test_spike_train_synchrony.py:
```diff
@@ -190,9 +192,255 @@ def test_spike_contrast_with_Izhikevich_network_auto(self):
+        block = neo.Block()
+
+        channel_index = neo.ChannelIndex(name='Channel 1', index=1)
+        block.channel_indexes.append(channel_index)
```
There's also still channelIndexes here...

dizcza replied:
right, we should get rid of old neo representation support in this PR

Kleinjohann replied:
Fixed.

dizcza commented on elephant/statistics.py:
```diff
@@ -974,23 +977,414 @@ def complexity_pdf(spiketrains, bin_size):
+            left_edges -= bin_shift
+
+            # ensure that an epoch does not start before the minimum t_start
+            min_t_start = min(st.t_start for st in self.input_spiketrains)
```
Why do you search for the smallest t_start when all the input spike trains are checked to have the same t_start in the `__init__` constructor function?
I'll replace it with `self.t_start`. Note: I already did this in `_epoch_with_spread` function:
before
```python
min_t_start = min(st.t_start for st in self.input_spiketrains)
left_edges[0] = max(min_t_start, left_edges[0])
```
after:
```python
left_edges[0] = max(self.t_start, left_edges[0])
```

Kleinjohann replied:
Thanks for spotting and fixing this, seems to be a remnant of the function we started from which did not require starts and stops to be aligned.

Kleinjohann commented:
I have fixed everything we discussed in our meeting on this and addressed all comments in here.

I spotted some double backticks in docstrings, are those a problem?

dizcza commented:
Thanks a lot!

> I spotted some double backticks in docstrings, are those a problem?

No, it's not a problem. Double vs single backticks is a matter of choice.

I'll merge it as soon as the tests pass.

P.S. Consider also adding into viziphant some of the nice synchroplots that you showed to us a month ago.
